### PR TITLE
Implement jar discovery and starting logic

### PR DESCRIPTION
**Pull request by weiiwang01**

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -18,6 +22,63 @@ class SpringBootCharm(CharmBase):
...
         super().__init__(*args)
+        self.framework.observe(self.on.config_changed, self.reconciliation)
+
+    def _spring_boot_container(self) -> ops.model.Container:
+        """Retrieve the container for the Spring Boot application.
+
+        Returns:
+            An instance of :class:`ops.charm.Container` represents the Spring Boot container.
+        """
+        container = self.unit.get_container("spring-boot-app")
```
Comment (arturo-seijas) [inline src/charm.py:33]: Why not returning it directly instead of assigning it to a variable?

File: src/charm.py [hunk 2]
```diff
@@ -18,6 +22,63 @@ class SpringBootCharm(CharmBase):
...
+
+    def _generate_spring_boot_layer(self) -> dict:
+        """Generate Spring Boot service layer for pebble.
+
+        Returns:
+            Spring Boot service layer configuration, in the form of a dict
+        """
+        container = self._spring_boot_container()
+        if not container.isdir("/app"):
+            raise ReconciliationError(new_status=BlockedStatus("dir /app does not exist"))
```
Comment (gregory-schiano) [inline src/charm.py:44]: Should we capitalize the message ?
Like "Directory /app doesn't exist"

File: src/charm.py [hunk 3]
```diff
@@ -18,6 +22,63 @@ class SpringBootCharm(CharmBase):
...
+        files_in_app = container.list_files("/app")
+        jar_files = [file.name for file in files_in_app if file.name.endswith(".jar")]
+        if not jar_files:
+            raise ReconciliationError(new_status=BlockedStatus("no jar file exists in /app"))
```
Comment (gregory-schiano) [inline src/charm.py:48]: Same as above: "No jar file found in /app"

File: src/charm.py [hunk 4]
```diff
@@ -18,6 +22,63 @@ class SpringBootCharm(CharmBase):
...
+        if len(jar_files) > 1:
+            raise ReconciliationError(new_status=BlockedStatus("multiple jar files exist in /app"))
```
Comment (gregory-schiano) [inline src/charm.py:50]: Also: "Multiple jar files found in /app"

File: src/charm.py [hunk 5]
```diff
@@ -18,6 +22,63 @@ class SpringBootCharm(CharmBase):
...
+        jar_file = jar_files[0]
+        return {
+            "services": {
```
Comment (gregory-schiano) [inline src/charm.py:53]: Pebble check should be added too

File: src/charm.py [hunk 6]
```diff
@@ -18,6 +22,63 @@ class SpringBootCharm(CharmBase):
...
+        if not container.isdir("/app"):
+            raise ReconciliationError(new_status=BlockedStatus("dir /app does not exist"))
+        files_in_app = container.list_files("/app")
+        jar_files = [file.name for file in files_in_app if file.name.endswith(".jar")]
+        if not jar_files:
+            raise ReconciliationError(new_status=BlockedStatus("no jar file exists in /app"))
+        if len(jar_files) > 1:
+            raise ReconciliationError(new_status=BlockedStatus("multiple jar files exist in /app"))
+        jar_file = jar_files[0]
+        return {
```
Comment (gregory-schiano) [inline src/charm.py:52]: I think the business logic of detecting the type of "jar" and output of the pebble plan should be split in 2 methods like:
1) detect_jar_type
2) create_peblle_plan

File: src/charm.py (hunk 5 above)
Comment (weiiwang01) [reply src/charm.py:53]: Added, thanks!

File: src/charm.py (hunk 1 above)
Comment (weiiwang01) [reply src/charm.py]: Fixed, thanks!

File: src/charm.py (hunk 6 above)
Comment (weiiwang01) [reply src/charm.py:65]: I separated the jar detection logic into a separate method `_detect_java_application`. Also, create abstractions for different types of java applications in `src/java_application.py`.

**Review by arturo-seijas — DISMISSED** [review]
LGTM

File: src/charm.py (hunk 3 above)
Comment (weiiwang01) [reply src/charm.py]: Updated, thanks!

File: src/charm.py (hunk 2 above)
Comment (weiiwang01) [reply src/charm.py]: Updated, thanks!

File: src/charm.py (hunk 4 above)
Comment (weiiwang01) [reply src/charm.py]: Updated, thanks!

File: src/charm.py [hunk 7]
```diff
@@ -18,6 +23,82 @@ class SpringBootCharm(CharmBase):
     def __init__(self, *args):
         """Initialize the instance."""
         super().__init__(*args)
+        self.framework.observe(self.on.config_changed, self.reconciliation)
+
+    def _detect_java_application(self) -> ExecutableJarApplication:
+        """Detect the type of the Java application inside the Spring Boot application image."""
```
Comment (amandahla) [inline src/charm.py:29]: Returns statement in docstring?

File: src/charm.py [hunk 8]
```diff
@@ -18,6 +23,82 @@ class SpringBootCharm(CharmBase):
...
+        container = self._spring_boot_container()
+        if container.isdir("/app"):
```
Comment (amandahla) [inline src/charm.py:31]: Do you think would be worth creating a constant for "/app"?

File: tests/unit/test_charm.py [hunk 9]
```diff
@@ -1,12 +1,81 @@
...
+
 """Spring Boot charm unit tests."""
+import ops.charm
+import pytest
 
+import exceptions
 
-import pytest
+
+def test_sprint_boot_pebble_layer(harness, patch):
```
Comment (amandahla) [inline tests/unit/test_charm.py:13]: What if /app doesn't exist?

File: src/charm.py (hunk 8 above)
Comment (gregory-schiano) [reply src/charm.py:31]: This will later be replaced by a charm config, so for the first iteration that's fine for me

File: src/charm.py (hunk 2 above)
Comment (gregory-schiano) [reply src/charm.py]: Let's see the outcome of your pool :)

File: src/charm.py [hunk 10]
```diff
@@ -18,6 +23,82 @@ class SpringBootCharm(CharmBase):
...
+                    "summary": "Spring Boot application service",
+                    "command": command,
+                    "startup": "enabled",
+                }
+            },
+            "checks": {
+                "wordpress-ready": {
+                    "override": "replace",
+                    "level": "alive",
+                    "http": {"url": "http://localhost/actuator/health"},
```
Comment (gregory-schiano) [inline src/charm.py:78]: Hmm I think the default port is 8080, so this shouldn't work I think

File: src/charm.py (hunk 7 above)
Comment (weiiwang01) [reply src/charm.py]: Updated, thanks!

File: tests/unit/test_charm.py (hunk 9 above)
Comment (weiiwang01) [reply tests/unit/test_charm.py:13]: The charm will enter blocking status with status message `Unknown Java application type`

File: src/charm.py (hunk 10 above)
Comment (weiiwang01) [reply src/charm.py]: Fixed, thanks!

**Review by arturo-seijas — APPROVED** [review]
LGTM

**Review by gregory-schiano — APPROVED** [review]
LGTM

**Pull request merged**